shimmy/meltingpot_compatibility.py

```diff
@@ -0,0 +1,193 @@
+"""Wrapper to convert a meltingpot substrate into a pettingzoo compatible environment
+
+Taken from
+https://github.com/deepmind/meltingpot/blob/main/examples/pettingzoo/utils.py
+and modified to modern pettingzoo API
+"""
+from __future__ import annotations
+
+import functools
+from typing import Tuple, Dict, Optional
+
+import gymnasium
+import numpy as np
+from gymnasium.utils import EzPickle
+import matplotlib.pyplot as plt
+from ml_collections import config_dict
+from pettingzoo.utils.env import ParallelEnv, AgentID
+from pettingzoo.utils.env import ObsDict, ActionDict
+
+import shimmy.utils.meltingpot as utils
+from meltingpot.python import substrate
+
+
+class MeltingPotCompatibilityV0(ParallelEnv, EzPickle):
+    """This compatibility wrapper converts a meltingpot substrate into a pettingzoo environment.
+
+    Melting Pot is a research tool developed to facilitate work on multi-agent artificial intelligence.
+    It assesses generalization to novel social situations involving both familiar and unfamiliar individuals,
+    and has been designed to test a broad range of social interactions such as: cooperation, competition,
+    deception, reciprocation, trust, stubbornness and so on.
+    Melting Pot offers researchers a set of over 50 multi-agent reinforcement learning substrates (multi-agent games)
+    on which to train agents, and over 256 unique test scenarios on which to evaluate these trained agents.
+    """
+
+    metadata = {'render_modes': ['human', 'rgb_array']}
+
+    PLAYER_STR_FORMAT = 'player_{index}'
+    MAX_CYCLES = 1000
+
+    def __init__(
+            self,
+            substrate_name: str,
+            render_mode: str | None,
+            max_cycles: int = MAX_CYCLES
+    ):
+        """Wrapper that converts a openspiel environment into a pettingzoo environment.
+
+        Args:
```

Comment:
Is it too complex to add the capability to pass an environment if someone has made a custom environment? Otherwise, don't worry

Reply:
I can look into that, they have a nice tutorial for making your own env so it would be nice to have.

Reply:
To update this, I was referring to this tutorial they have on making custom envs: https://github.com/deepmind/meltingpot/blob/main/docs/substrate_tutorial/index.md
It's not clear how exactly you use an environment besides loading it explicitly through an import statement, and their link to an example game is dead so I can't look at that for reference. 

I think the best option for this kind of functionality would be to make something like `custom_environment_compatibility` which takes an already defined env (which you import directly at the top of the file, rather than passing substrate name to be built by a factory). I tried to implement it as you described with taking in `env` in the constructor, but either the `substrate_name` or `env` has to be a required arg, and I'm not sure how to make args mutually exclusive/only have one of the two required. I guess we could just have it be that you pass in a substrate_name anyways which isn't used, if you're loading from an existing env.

Here's my rough code idea for what you described:
```
    def __init__(
        self,
        substrate_name: str,
        render_mode: str | None,
        max_cycles: int = MAX_CYCLES,
        env: meltingpot.python.utils.substrates.substrate.Substrate = None,
    ):
        """Wrapper that converts a openspiel environment into a pettingzoo environment.

        Args:
            substrate_name (str): name of meltingpot substrate to load
            render_mode (Optional[str]): render_mode
            max_cycles (Optional[int]): maximum number of cycles (steps) before termination
            env (Optional[meltingpot.python.utils.substrates.substrate.Substrate]): existing meltingpot env to use

        """
        EzPickle.__init__(self, substrate_name, render_mode, max_cycles, env)

        if env is not None:
            self._env = env
        else:
            # Create env config
            self.substrate_name = substrate_name
            self.player_roles = meltingpot.python.substrate.get_config(
                self.substrate_name
            ).default_player_roles
            self.max_cycles = max_cycles
            self.env_config = {
                "substrate": self.substrate_name,
                "roles": self.player_roles,
            }

            # Build substrate from pickle
            self.env_config = config_dict.ConfigDict(self.env_config)
            self._env = meltingpot.python.substrate.build(
                self.env_config["substrate"], roles=self.env_config["roles"]
            )
```